ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -1118,6 +1085,123 @@ module ClosestKdc = struct
       Xapi_periodic_scheduler.remove_from_queue periodic_update_task_name
 end
 
+module RotateMachinePassword = struct
+  let task_name = "Rotating machine password"
+
+  let tmp_krb5_conf =
+    Printf.sprintf "%s/pwrotate.krb5" (Filename.get_temp_dir_name ())
+
+  let kdc_fqdn_of_ip kdc =
+    let args =
+      ["ads"; "lookup"; "--server"; kdc; "--kerberos"; "-d"; debug_level ()]
+    in
+    Helpers.call_script !Xapi_globs.net_cmd args ~log_output:On_failure
+    |> Xapi_cmd_result.of_output ~sep:':' ~key:"Domain Controller"
+
+  let generate_krb5_tmp_config ~domain ~kdc_fqdn =
+    (* Configure which server to change the password
+     * https://web.mit.edu/kerberos/krb5-devel/doc/admin/conf_files/krb5_conf.html *)
+    let open Xapi_stdext_unix in
+    let realm = String.uppercase_ascii domain in
+    let domain_netbios =
+      Wbinfo.domain_name_of ~target_name_type:NetbiosName ~from_name:domain
+      |> Result.get_ok
+    in
+    let winbind_krb5 = Ldap.krb5_conf_path ~domain_netbios in
+    let include_item =
+      if Sys.file_exists winbind_krb5 then
+        [Printf.sprintf "include %s" winbind_krb5]
+      else
+        []
+    in
+
+    let conf_contents =
+      String.concat "\n"
+        ([
+           "# auto-generated by xapi"
+         ; "[libdefaults]"
+         ; Printf.sprintf "default_realm = %s" realm
+         ; "[realms]"
+         ; Printf.sprintf "%s={" realm
+         ; Printf.sprintf "kpasswd_server=%s" kdc_fqdn
+         ; Printf.sprintf "kdc=%s" kdc_fqdn
+         ; "}"
+           (* include winbind generated configure if exists *)
+         ]
+        @ include_item
+        @ [""] (* Empty line at the end *)
+        )
+    in
+
+    let len = String.length conf_contents in
+    Unixext.atomic_write_to_file tmp_krb5_conf 0o0644 (fun fd ->
+        let (_ : int) = Unix.single_write_substring fd conf_contents 0 len in
+        Unix.fsync fd
+    )
+
+  let clear_tmp_krb5_conf () =
+    if !Xapi_globs.winbind_keep_configuration then
+      ()
+    else
+      try Sys.remove tmp_krb5_conf
+      with e ->
+        debug "Failed to remove tmp keytab file  %s" (ExnHelper.string_of_exn e)
+
+  let rotate () =
+    let now = Unix.time () in
+    let now_str = string_of_float now in
+    try
+      let {service_name; machine_pwd_last_change_time} =
+        get_domain_info_from_db ()
+      in
+      match machine_pwd_last_change_time with
+      | Some time when now < time +. !Xapi_globs.winbind_machine_pwd_timeout ->
```

Comment:
This will need some testing, but there might be some rounding issue here where we'll always miss the first call to rotate (it is just a bit too soon) and the interval between password rotations will be sometimes machine_password_timeout, sometimes twice that. Probably doesn't matter in practice though. (the scheduler and this code also use different clocks, monotonic vs wall clock)

Reply:
this can be fix by like update the periodic interval to winbind_machine_pwd_timeout/2, 
but as you said, this does not matter in practice as rotating machine password is not in the critical path.